[SUREFIRE-1152] Support rerunFailingTestsCount for test suites

seanf opened the pull request:
Submitted by: Sean Flanigan

Tibor17 commented:
@seanf 
I need to find more time, but I will find in several days, but the problem I have is that you removed adding methods to the failures list. Shouldn't we distinguish between suite and non-suite runner in order to add class or method to the list, respectively? Are all ITs passed in your build?

seanf commented:
I don't think I did remove adding methods to the failure list.  I just changed it from adding the method name (and throwing away the class name) to adding the `ClassMethod`, which includes both class name and method name.  So now when executing a test method, we know which class it came from, whether it was found via a suite or not.

All the integration tests I am able to run are still passing, although I found that one of them was running out of memory, which prevents the build from progressing any further.  But even before that, I wasn't able to run everything.  This unit test in another module (`surefire-junit47`) is failing, even without my changes:

    beforeAfterThreadChanges(org.apache.maven.surefire.junitcore.pc.ParallelComputerBuilderTest): (..)

I'll update this comment with the out-of-memory test when it happens again.

seanf commented:
Is it just me, or do the messages like 

    Running org.apache.maven.surefire.its.jiras.Surefire408ManualProviderSelectionIT

actually appear when the test has *finished*?  Makes it hard to tell why the build is sitting there.

Tibor17 commented:
export M2_HOME to apache-maven-3.2.3
export MAVEN_OPTS a -server -Xmx768m -XX:MaxPermSize=512m
The heap can be a bit larger, but not too much. This setting works on CI and my machine as well.
Please run $mvn install -Prun-its
and do not run anything on your PC meanwhile. It executes performance tests, so please no IO activity, no CPU activity for 40 minutes, just let the tests running.

Tibor17 commented:
`surefire-junit47` performs performance test. The CPU goes up down for one minute and there do not do anything. Then some IT run parallel tests and it needs CPU again, but it is not so critical, e.g. JUniParallelIT takes 5 min and the CPU bumps again, so I know that the middle IT time executes this IT and I just take a coffee.

seanf commented:
Okay, thanks, I'll have a look at my pc in 40 minutes then.
On 11/04/2015 8:54 pm, "Tibor Digana" <notifications@github.com> wrote:

> export M2_HOME to apache-maven-3.2.3
> export MAVEN_OPTS a -server -Xmx768m -XX:MaxPermSize=512m
> The heap can be a bit larger, but not too much. This setting works on CI
> and my machine as well.
> Please run $mvn install -Prun-its
> and do not run anything on your PC meanwhile. It executes performance
> tests, so please no IO activity, no CPU activity for 40 minutes, just let
> the tests running.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/maven-surefire/pull/89#issuecomment-91825428>.
>

Tibor17 commented:
Depends on your CPU. For instance CI executes within 23min or 19 min, but my PC is slower so it takes cca 34 min depends.

seanf commented:
Well, it helped, but I still had three failures on master branch (27 minutes):

	Tests in error: 
	  jarShouldUseFile(org.apache.maven.surefire.its.jiras.Surefire855AllowFailsafeUseArtifactFileIT): Exit code was non-zero: 1; command line and log = (..)
	  warShouldUseClasses(org.apache.maven.surefire.its.jiras.Surefire855AllowFailsafeUseArtifactFileIT): Exit code was non-zero: 1; command line and log = (..)
	  osgiBundleShouldUseFile(org.apache.maven.surefire.its.jiras.Surefire855AllowFailsafeUseArtifactFileIT): Exit code was non-zero: 1; command line and log = (..)

I'll ignore those and try again with my branch.

seanf commented:
Same again on my branch, but with one more passing test (ie Surefire1152RerunFailingTestsInSuiteIT).

Tibor17 commented:
This means you broke backwards compatibility because the failsafe plugin should run JAR binary instead of test-classes, that's about Surefire855. The exception is war, and other packaging.

We should test your line changes and retesting every change step by step, but you can do that same.

seanf commented:
As I said, this test was already broken in master.  Even without my
changes.

On Monday, April 13, 2015, Tibor Digana <notifications@github.com> wrote:

> This means you broke backwards compatibility because the failsafe plugin
> should run JAR binary instead of test-classes, that's about Surefire855.
> The exception is war, and other packaging.
>
> We should test your line changes and retesting every change step by step,
> but you can do that same.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/maven-surefire/pull/89#issuecomment-92099752>.
>

Tibor17 commented:
That's strange because Surefire855 passed on CI machine
https://builds.apache.org/job/maven-surefire-mvn-2.2.1/org.apache.maven.surefire$surefire-integration-tests/1579/testReport/org.apache.maven.surefire.its.jiras/Surefire855AllowFailsafeUseArtifactFileIT/

I will run just only this test on my PC.

seanf commented:
I've made the changes you requested, except I haven't changed the assertions in `Surefire1152RerunFailingTestsInSuiteIT`, because I need an answer from you about testing surefire and/or failsafe for this bug.  

My review changes are in a separate commit for easier review, so feel free to use squash merge.

seanf commented:
The only non-private class change should be this change in
JUnit4ProviderUtil.java:

    -    public static Set<String> generateFailingTests( List<Failure>
allFailures )
    +    public static Set<ClassMethod> generateFailingTests( List<Failure>
allFailures )

I could make it two different methods if it will help, but I think the old
method signature was flawed, because a method name String is not enough to
identify a test.

That's assuming JUnit4ProviderUtil is considered part of the API, which I
hope it is not.  And if it isn't, I don't see how this patch could affect
the API.


On 13 April 2015 at 08:57, Tibor Digana <notifications@github.com> wrote:

> That's strange because Surefire855 passed on CI machine
>
> https://builds.apache.org/job/maven-surefire-mvn-2.2.1/org.apache.maven.surefire$surefire-integration-tests/1579/testReport/org.apache.maven.surefire.its.jiras/Surefire855AllowFailsafeUseArtifactFileIT/
>
> I will run just only this test on my PC.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/maven-surefire/pull/89#issuecomment-92147353>.
>

seanf commented:
@Tibor17 Are you waiting for something from me on this pull request?  Thanks.

Tibor17 commented:
@seanf 
No, not yet. I was only busy this week. I will have a look because this fix in 2.19 is important in such new feature like rerunFailingTestsCount to become stable.

Tibor17 commented:
`Surefire855AllowFailsafeUseArtifactFileIT` is the one which still fails in your project?
In my project it passed but of course I launched the single test.

Did you test the Suite with `surefire-junit47` provider in your IT as well?
If not, we should add some Maven profile with switching the profiles.

In next step I will debug the test, then we will make a small change and we are through.

seanf commented:
> `Surefire855AllowFailsafeUseArtifactFileIT` is the one which still fails in your project?
In my project it passed but of course I launched the single test.

Yes, that's the only test class which has been failing, both in master and in my branch.

> Did you test the Suite with `surefire-junit47` provider in your IT as well?
If not, we should add some Maven profile with switching the providers.

I'm not sure what you mean.  `Surefire855AllowFailsafeUseArtifactFileIT` failed when I ran the entire test suite the way you recommended.  It also fails when run by itself like this:

    mvn install -Prun-its -Dit.test=Surefire855AllowFailsafeUseArtifactFileIT

> In next step I will debug the test, then we will make a small change and we are through.

Have you tried testing my change using your environment (or the CI server)?  Whatever is wrong with `Surefire855AllowFailsafeUseArtifactFileIT` seems to depend on the environment.

Tibor17 commented:
The CI tests passed in `Surefire855AllowFailsafeUseArtifactFileIT `
https://builds.apache.org/job/maven-surefire/
If you have log report from this test, send it to me by email and I will have a look.

The `surefire-junit47` is not related to `Surefire855AllowFailsafeUseArtifactFileIT`. As far as you know the default provider is `surefire-junit4` which works with JUnit4+, but we have also newer provider able to work with JUni4.7+ enabling features (e.g. parallel execution) which were not introduced in JUnit 4.0. Therefore we have to backport the fixes to `surefire-junit47`. Currently your fix is placed in `surefire-junit4` module.
So I would like to kindly ask you to extend your IT testes with both providers. Our documentatio page demonstrates on how the providers can be selected 
http://maven.apache.org/surefire/maven-surefire-plugin/examples/providers.html
This IT is a good example which may help you 
https://github.com/seanf/maven-surefire/blob/master/surefire-integration-tests/src/test/resources/junit48-method-pattern/pom.xml

If you have any questions, don't  hesitate to ask.

seanf commented:
On 28 April 2015 at 05:30, Tibor Digana <notifications@github.com> wrote:

> The CI tests passed in Surefire855AllowFailsafeUseArtifactFileIT
> https://builds.apache.org/job/maven-surefire/
>
If you have log report from this test, send it to me by email and I will
> have a look.
>

Okay, will do.


> The surefire-junit47 is not related to
> Surefire855AllowFailsafeUseArtifactFileIT. As far as you know the default
> provider is surefire-junit4 which works with JUnit4+, but we have also
> newer provider able to work with JUni4.7+ enabling features (e.g. parallel
> execution) which were not introduced in JUnit 4.0. Therefore we have to
> backport the fixes to surefire-junit47. Currently your fix is placed in
> surefire-junit4 module.
> So I would like to kindly ask you to extend your IT testes with both
> providers. Our documentatio page demonstrates on how the providers can be
> selected
>
> http://maven.apache.org/surefire/maven-surefire-plugin/examples/providers.html
> This IT is a good example which may help you
>
> https://github.com/seanf/maven-surefire/blob/master/surefire-integration-tests/src/test/resources/junit48-method-pattern/pom.xml
>

Thanks, I'll see what I can do.

> If you have any questions, don't hesitate to ask.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/maven-surefire/pull/89#issuecomment-96789238>.
>

seanf commented:
@Tibor17 Okay, I have added a profile for testing `surefire-junit47` and changed `JUnit4ProviderUtil` so that it shares more code between the two providers, thus fixing `rerunFailingTestsCount` for suites on surefire-junit47.

Tibor17 commented:
The test `testJUnit48Provider47` failed in my project.
After I have removed the intersection of filters JUnitCoreProvider at L155 and used only `failingMethodsFilter`, I got success:
`Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Flakes: 3` in both child ITs.

Anyway the entire build failed with `Surefire1152RerunFailingTestsInSuiteIT` I guess because of the assertion statement at L47 in `Surefire1152RerunFailingTestsInSuiteIT`.

I guess the filter intersection, the surefire originally had, of `<include>FlakyITSuite</include>` and `FlakyTest` was unnecessary.

Would you make your local changes and build all project to see if we break other ITs?

seanf commented:
On 29 April 2015 at 07:26, Tibor Digana <notifications@github.com> wrote:

> The test testJUnit48Provider47 failed in my project.
> After I have removed the intersection of filters JUnitCoreProvider at L155
> and used only failingMethodsFilter, I got success:
> Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Flakes: 3 in both child
> ITs.
>
> Anyway the entire build failed with Surefire1152RerunFailingTestsInSuiteIT
> I guess because of the assertion statement at L47 in
> Surefire1152RerunFailingTestsInSuiteIT.
>

Oh right, when I added the junit4/junit47 profiles, I accidentally switched
from executeVerify() to executeTest(), so it was no longer running FlakyIT
through failsafe.

By the way, I never saw a reply to my question regarding surefire vs
failsafe:
https://github.com/seanf/maven-surefire/commit/cb977633f0f1d2be4b2556a4743a5f52d8a7008a#commitcomment-10685280

Should I remove *all* my failsafe/verify/FlakyIT stuff, and test the
providers via surefire only?

> I guess the filter intersection, the surefire originally had, of
> <include>FlakyITSuite</include> and FlakyTest was unnecessary.
>
Yes, I suspected it was a redundant no-op, so I'm surprised eliminating it
changed anything for you.  I've taken it out now.

> Would you make your local changes and build all project to see if we break
> other ITs?
>
Okay, I've made the changes.

I tried running the full build, but it got stuck after this:

Running
org.apache.maven.surefire.its.jiras.Surefire570MultipleReportDirectoriesIT
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 17.144 sec

Running the test project does work, at least when run like this:

    mvn verify -f
surefire-integration-tests/src/test/resources/surefire-1152-rerunFailingTestsCount-suite/pom.xml
-Dsurefire.version=2.19-SNAPSHOT -P surefire-junit4
    mvn verify -f
surefire-integration-tests/src/test/resources/surefire-1152-rerunFailingTestsCount-suite/pom.xml
-Dsurefire.version=2.19-SNAPSHOT -P surefire-junit47

But when I tried to run Surefire1152RerunFailingTestsInSuiteIT by itself
(after mvn installing the provider modules) like this -

    mvn verify -pl surefire-integration-tests
-Dit.test=Surefire1152RerunFailingTestsInSuiteIT

- I got initialiser errors (classpath problems?) and/or weird non-SNAPSHOT
versions of surefire artifacts which has made debugging impossible so far.
I'm not sure if that's the right way to run a single test though.

I'm running the full build again, but I won't see the results until next
week.

Tibor17 commented:
:) Okay I know your situation. If I am implementing really a big feature a lot tests fail 30 - 150 tests, but that's pretty normal as long as the feature development starts.
In order to run a single test I am usually changing `surefire-integration-tests/pom.xml`.
You can see test inclusion `*IT*` which can be substituted by your test `Surefire1152RerunFailingTestsInSuiteIT`.

Tibor17 commented:
LGTM

I would merge your branch with master. Do you plan new changes in the code?

seanf commented:
Assuming the full build is green, then I'm happy. Go ahead, thanks @Tibor17.

Tibor17 commented:
Thank you for contributing.
